crates/virtio-queue/src/lib.rs

```diff
@@ -384,11 +385,9 @@ impl VirtqUsedElem {
 
 unsafe impl ByteValued for VirtqUsedElem {}
 
+/// Struct to maintain information and manipulate state of a virtio queue.
 #[derive(Clone, Debug)]
-/// A virtio queue's parameters.
-pub struct Queue<M: GuestAddressSpace> {
-    mem: M,
-
+pub struct QueueState<M: GuestAddressSpace> {
```

Comment:
I have tried to follow the suggestion, but the result is neutral, the `QueueState` has been simplified by removing the type parameter, the callers of QueueState's. All QueueState's methods take a `mem: &GuestAddressSpace::M`, so when calling those methods, you need to `queuestate.avail_idx::<M>(&self.mem.memory(), order)`.
So I suggestion to keep the type parameter the ease the method invoke.

Reply:
Seems we also need <M: GuestAddressSpace> for
```
impl<M: GuestAddressSpace> QueueStateT<M> for QueueState<M>
```

Reply:
Hmm, I think that can be worked around as well; I'll quickly try to better illustrate what I have in mind by adding an example commit on top of your changes in a branch somewhere, and ping here with a link when it's done a bit later.

Reply:
I experimented some more with this, and I pushed an example to [this](https://github.com/alexandruag/vm-virtio/tree/temp4) branch. The last two commits are placed on top of the changes you introduce in this PR, and we can just take and use them if they make sense. Can you please have a look and share your thoughts? I'll also propose a small change to how we offer the `lock` semantics that you mentioned are required already; will ping back soon about that.

Reply:
Btw I just found out why that weird `Sized` trait bound had to appear in the above example; it's because of the `impl Bytes for T where T: GuestMemory` auto implementation from `vm-memory`, which implies a `Sized` bound for `T`. I think we'll have to get rid of the auto impl at some point because it prevents a couple of useful scenarios, but for now I'll add a small change to relax the bounds there to include `?Sized` as well (since there's no reason really not to).

Reply:
> I experimented some more with this, and I pushed an example to [this](https://github.com/alexandruag/vm-virtio/tree/temp4) branch. The last two commits are placed on top of the changes you introduce in this PR, and we can just take and use them if they make sense. Can you please have a look and share your thoughts? I'll also propose a small change to how we offer the `lock` semantics that you mentioned are required already; will ping back soon about that.

Thanks for your work to simplify the interface, it looks attractive. One drawback with the simplification is that we can't make trait object of `QueueStateT` anymore. I'm not sure whether there's any case needing `QueueStateT` trait object.